File: DSA Busted/Heap/MergeKSortedLists.cpp

// https://leetcode.com/problems/merge-k-sorted-lists/description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    class compare{
        public:
            bool operator()(ListNode* node1, ListNode* node2){
                return node1->val>node2->val;
            }
    };
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.size()==0)
            return NULL;
        priority_queue<ListNode*, vector<ListNode*>, compare> minHeap;
        for(auto i : lists)
            if(i!=NULL)
                minHeap.push(i);
        
        ListNode* head= NULL;
        ListNode* tail= NULL;

        while(!minHeap.empty()){
            ListNode* top= minHeap.top();
            minHeap.pop();

            if(top->next){
                minHeap.push(top->next);
            }

            if(head==NULL){
                head= top;
                tail= top;
            }else{
                tail->next= top;
                tail= top;
            }
        }
        return head;
    }
};

File: DSA-Problems/Heap/MergeKSortedLists.java
import java.util.*;

// https://leetcode.com/problems/merge-k-sorted-lists/description/

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

public class MergeKSortedLists {
    class Compare implements Comparator<ListNode> {
        @Override
        public int compare(ListNode a, ListNode b) {
            return a.val - b.val;
        }
    }

    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>(new Compare());

        for (ListNode node : lists) {
            if (node != null) pq.offer(node);
        }

        ListNode dummy = new ListNode(0);
        ListNode current = dummy;

        while (!pq.isEmpty()) {
            ListNode top = pq.poll();
            current.next = top;
            current = current.next;
            if (top.next != null) pq.offer(top.next);
        }

        return dummy.next;
    }
}